--- ../asuswrt/release/src/router/rc/common.c.orig
+++ ../asuswrt/release/src/router/rc/common.c
@@ -1416,3 +1416,28 @@
 }
 #endif
 
+void run_custom_script(char *name)
+{
+	char script[120];
+
+	sprintf(script, "/jffs/scripts/%s", name);
+
+	if(f_exists(script)) {
+		_dprintf("Script: running %s\n", script);
+		xstart(script);
+	}
+}
+
+void run_custom_script_blocking(char *name, char *args)
+{
+	char script[120];
+
+	sprintf(script, "/jffs/scripts/%s", name);
+
+	if(f_exists(script)) {
+		_dprintf("Script: running %s\n", script);
+		eval(script, args);
+	}
+
+}
+
--- ../asuswrt/release/src/router/rc/firewall.c.orig
+++ ../asuswrt/release/src/router/rc/firewall.c
@@ -4092,6 +4092,8 @@
 	}
 #endif
 
+	run_custom_script("firewall-start");
+
 	return 0;
 }
 
--- ../asuswrt/release/src/router/rc/services.c.orig
+++ ../asuswrt/release/src/router/rc/services.c
@@ -2475,6 +2475,8 @@
 	start_webdav();
 #endif
 
+	run_custom_script("services-start");
+
 	return 0;
 }
 
@@ -2491,6 +2493,8 @@
 stop_services(void)
 {
 
+	run_custom_script("services-stop");
+
 #ifdef RTCONFIG_WEBDAV
 	stop_webdav();
 #endif
--- ../asuswrt/release/src/router/rc/wan.c.orig
+++ ../asuswrt/release/src/router/rc/wan.c
@@ -793,6 +793,9 @@
 	else if(state == WAN_STATE_STOPPING){
 		unlink("/tmp/wanstatus.log");
 	}
+        else if (state == WAN_STATE_CONNECTED) {
+                run_custom_script("wan-start");
+        }
 }
 
 #ifdef RTCONFIG_IPV6
--- ../asuswrt/release/src/router/rc/init.c.orig
+++ ../asuswrt/release/src/router/rc/init.c
@@ -1886,6 +1886,8 @@
 	start_jffs2();
 #endif
 
+	run_custom_script("init-start");
+
 	state = SIGUSR2;	/* START */
 
 	for (;;) {
--- ../asuswrt/release/src/router/rc/usb.c.orig
+++ ../asuswrt/release/src/router/rc/usb.c
@@ -843,6 +843,8 @@
 	if(!is_valid_hostname(the_label))
 		memset(the_label, 0, 128);
 
+	run_custom_script_blocking("pre-mount", dev_name);
+
 	if (f_exists("/etc/fstab")) {
 		if (strcmp(type, "swap") == 0) {
 			_eval(swp_argv, NULL, 0, NULL);
@@ -980,6 +982,7 @@
 
 		if (nvram_get_int("usb_automount"))
 			run_nvscript("script_usbmount", mountpoint, 3);
+		run_custom_script_blocking("post-mount", mountpoint);
 
 #if defined(RTCONFIG_APP_PREINSTALLED) && defined(RTCONFIG_CLOUDSYNC)
 		char word[PATH_MAX], *next_word;
