--- ../asuswrt/release/src/router/rc/common.c.orig	2012-09-07 10:57:08.000000000 +0400
+++ ../asuswrt/release/src/router/rc/common.c	2012-09-13 23:31:11.764683061 +0400
@@ -1413,3 +1413,28 @@
 }
 #endif
 
+void run_custom_script(char *name)
+{
+	char script[120];
+
+	sprintf(script, "/jffs/scripts/%s", name);
+
+	if(f_exists(script)) {
+		_dprintf("Script: running %s\n", script);
+		xstart(script);
+	}
+}
+
+void run_custom_script_blocking(char *name, char *args)
+{
+	char script[120];
+
+	sprintf(script, "/jffs/scripts/%s", name);
+
+	if(f_exists(script)) {
+		_dprintf("Script: running %s\n", script);
+		eval(script, args);
+	}
+
+}
+
--- ../asuswrt/release/src/router/rc/firewall.c.orig	2012-09-07 10:57:08.000000000 +0400
+++ ../asuswrt/release/src/router/rc/firewall.c	2012-09-13 23:31:11.768683062 +0400
@@ -4088,6 +4088,8 @@
 	}
 #endif
 
+	run_custom_script("firewall-start");
+
 	return 0;
 }
 
--- ../asuswrt/release/src/router/rc/services.c.orig	2012-09-07 10:57:08.000000000 +0400
+++ ../asuswrt/release/src/router/rc/services.c	2012-09-13 23:31:11.780683059 +0400
@@ -2470,6 +2470,8 @@
 	start_webdav();
 #endif
 
+	run_custom_script("services-start");
+
 	return 0;
 }
 
@@ -2486,6 +2488,8 @@
 stop_services(void)
 {
 
+	run_custom_script("services-stop");
+
 #ifdef RTCONFIG_WEBDAV
 	stop_webdav();
 #endif
--- ../asuswrt/release/src/router/rc/usb.c.orig	2012-09-07 10:57:08.000000000 +0400
+++ ../asuswrt/release/src/router/rc/usb.c	2012-09-13 23:31:11.796683062 +0400
@@ -838,6 +838,8 @@
 	if(!is_valid_hostname(the_label))
 		memset(the_label, 0, 128);
 
+	run_custom_script_blocking("pre-mount", dev_name);
+
 	if (f_exists("/etc/fstab")) {
 		if (strcmp(type, "swap") == 0) {
 			_eval(swp_argv, NULL, 0, NULL);
@@ -1064,6 +1066,7 @@
 			start_cloudsync();
 		}
 #endif
+		run_custom_script_blocking("post-mount", mountpoint);
 	}
 	return (ret == MOUNT_VAL_RONLY || ret == MOUNT_VAL_RW);
 }
--- ../asuswrt/release/src/router/rc/wan.c.orig	2012-09-07 10:57:08.000000000 +0400
+++ ../asuswrt/release/src/router/rc/wan.c	2012-09-13 23:31:11.796683063 +0400
@@ -745,6 +745,9 @@
 	else if(state == WAN_STATE_STOPPING){
 		unlink("/tmp/wanstatus.log");
 	}
+        else if (state == WAN_STATE_CONNECTED) {
+                run_custom_script("wan-start");
+        }
 }
 
 #ifdef RTCONFIG_IPV6
--- ../asuswrt/release/src/router/rc/init.c.orig	2012-09-07 10:57:08.000000000 +0400
+++ ../asuswrt/release/src/router/rc/init.c	2012-09-13 23:31:11.796683063 +0400
@@ -1872,6 +1872,8 @@
 	start_jffs2();
 #endif
 
+	run_custom_script("init-start");
+
 	state = SIGUSR2;	/* START */
 
 	for (;;) {
